packages/tx/src/eip2930Transaction.ts

```diff
@@ -0,0 +1,385 @@
+      this.data,
+      this.accessList,
+      this.v !== undefined ? bnToRlp(this.v) : Buffer.from([]),
+      this.r !== undefined ? bnToRlp(this.r) : Buffer.from([]),
+      this.s !== undefined ? bnToRlp(this.s) : Buffer.from([]),
+    ]
+    if (!asList) {
+      return Buffer.concat([Buffer.from('01', 'hex'), rlp.encode(base)])
+    } else {
+      return base
+    }
+  }
+
+  /**
+   * Returns the encoding of the transaction. For typed transaction, this is the raw Buffer.
+   * In LegacyTransaction, this is a Buffer array.
+   */
+  serialize(): Buffer {
+    return <Buffer>this.raw()
+  }
+
+  /**
+   * Returns an object with the JSON representation of the transaction
+   */
+  toJSON(): JsonTx {
+    const accessListJSON = []
+
+    for (let index = 0; index < this.accessList.length; index++) {
+      const item: any = this.accessList[index]
+      const JSONItem: any = {
+        address: '0x' + setLengthLeft(<Buffer>item[0], 20).toString('hex'),
+        storageKeys: [],
+      }
+      const storageSlots: Buffer[] = item[1]
+      for (let slot = 0; slot < storageSlots.length; slot++) {
+        const storageSlot = storageSlots[slot]
+        JSONItem.storageKeys.push('0x' + setLengthLeft(storageSlot, 32).toString('hex'))
+      }
+      accessListJSON.push(JSONItem)
+    }
+
+    return {
+      chainId: bnToHex(this.chainId),
+      nonce: bnToHex(this.nonce),
+      gasPrice: bnToHex(this.gasPrice),
+      gasLimit: bnToHex(this.gasLimit),
+      to: this.to !== undefined ? this.to.toString() : undefined,
+      value: bnToHex(this.value),
+      data: '0x' + this.data.toString('hex'),
+      accessList: accessListJSON,
+    }
+  }
+
+  public isSigned(): boolean {
+    const { yParity, r, s } = this
+    return yParity !== undefined && !!r && !!s
+  }
+
+  public hash(): Buffer {
+    if (!this.isSigned()) {
+      throw new Error('Cannot call hash method if transaction is not signed')
+    }
+
+    return keccak256(this.serialize())
+  }
+
+  public getMessageToVerifySignature(): Buffer {
+    return this.getMessageToSign()
+  }
+
+  public getSenderPublicKey(): Buffer {
+    if (!this.isSigned()) {
+      throw new Error('Cannot call this method if transaction is not signed')
+    }
+
+    const msgHash = this.getMessageToVerifySignature()
+
+    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
+    // TODO: verify if this is the case for EIP-2930
+    if (this.common.gteHardfork('homestead') && this.s && this.s.gt(N_DIV_2)) {
+      throw new Error(
+        'Invalid Signature: s-values greater than secp256k1n/2 are considered invalid'
+      )
+    }
+
+    const { yParity, r, s } = this
+    if (yParity === undefined || !r || !s) {
+      throw new Error('Missing values to derive sender public key from signed tx')
+    }
+
+    try {
+      return ecrecover(
+        msgHash,
+        yParity.toNumber() + 27, // Recover the 27 which was stripped from ecsign
```

Comment:
I am confused here, so is `yParity` an alias to `v` or does this have a difference of `27`?

Reply:
The EIP states that `yParity` is either `0` or `1`. `ecsign` internally adds 27 to the output of `secp256k1`. Thus when we get the signature's `v` value, we either get `27` or `28`, to get the right `v` value we subtract 27. But in order to use `ecrecover` (which expects a number >=27) we have to "recover" the original output value from `ecsign`.

Reply:
Ok, I've read this again a bit also in [EIP-155](https://eips.ethereum.org/EIPS/eip-155). So if I get correctly, both `v` (in legacy) and `yParity` (in EIP2930) are referrring to the `v` value (so are indeed aliases), it is just the case that different `v` values for the different transaction types are used.

If this is the case (correct me if I am wrong) I would pledge to:

1. Move

```typescript
public readonly v?: BN
public readonly r?: BN
public readonly s?: BN
```

to `BaseTransaction`.

2. Drop at least the `s` and `r` aliases in `EIP2930Transaction`, so `senderS()` and `senderR()`. If people uses these this will only lead to the situation that they are making their code incompatible to work with a `LegacyTransaction`, which is unnecessary and rather annoying in doubt.

3. Eventually also drop the `yParity()` alias, not sure. Might be somewhat nice to reflect the differences, but at the end same argumentation as in 2.

4. Move as much generic signature functionality as possible into `BaseTransaction`